README
	Paul Fox
	Foxtrot Systems
	Consultancy Services

	paul.d.fox@gmail.com

	The ctw Widget (Version 1.3) March 1992
	=======================================

This README file describes the 'ctw' (Color Terminal Widget) 
for X windows (X11R4 or later).

CTW is an emulation of an ANSI terminal compatable with the VTx00
series of terminals containing much of the same functionality that
xterm provides. Unfortunately, xterm does not support color and
is not cleanly written enough to allow any of the code to be reused
by other applications.

I wrote CTW because I like to use the color of my terminal, and in
particular I like to have yellow prompts with a green text for the
rest (so the prompt stands out a little). I also use CTW to run
my color editor -- CRISP -- within the terminal emulator rather than
having to spawn a new window.

Acknowledgements
================
I would like to thank the following people who have contributed
code, and or ideas to improve ctw. If I have omitted anyone please
mail me and I'll put you down here, and my apologies in advance.

Warren Tucker -- thanks for making me thank that allowing the colors
to be changed would be a good idea. Also other snippets as well.


Jim Wight (J.K.Wight@newcastle.ac.uk) -- thanks for making me believe
that handling of rows/columns and other issues to do with xterm 
compatability aren't that difficult to do properly, and examples
of code to improve CTW.

Platforms & Portability
=======================
I have only tested ctw on a Sun running SunOS 4.1. I compile
it under Open windows, and it runs quite happily under X11R5.
ctw should be pretty much machine independent (although it assumes
32-bit longs/ints). (It also uses the usleep() system call but you can
comment that out if you dont have it -- its not very important).

fcterm is an Athena Widgets based terminal emulator.
fcterm is not as portable. It works on the Sun, but things like 
finding a pty and spawning the child process does not handle all the
exceptions which xterm does handle. fcterm is a pretty small program at
present, so it shouldn't be too difficult to port -- especially if
you have the xterm sources handy. (fcterm may look daunting at first
but its just a hackers job. ctw is much cleaner).

fmcterm is a Motif version of the terminal emulator. The user interface
uses the menu bar to access options rather than a popup window in the
terminal area.

ctw & fcterm come with a simple makefile. If anybody wants to write
an Imakefile then please do so.

Distribution
============
I hereby place this software in the public domain in the hope that
it will further mankinds dependence on computers :-). People
may use this software in any commercial application without the
need for any further permission, except that I ask that my name
be acknowledged in your product.

It would be nice if the souls at MIT would like to incorporate ctw
into the core release, and if so please let me know what I need
to do to get it in a fit state for the core release. 

Overview
========
The idea behind CTW is to properly encapsulate all the functionality
of a widget which supports a public function (ctw_add_string) which is
called to add characters to the display in a similar way to printf().
One could take a program which is based on printf()/getchar() and with a
little amount of work, have the program run in its own X window as an
alternative to running from within an xterm. 

ctw is contained completely within the files:

	ctw.h	-- Public include file
	ctwP.h	-- Private widget definitions
	ctw.c	-- The implementation

In addition the following files are supplied:

	fcterm.c -- Athena based terminal emulator
	fmcterm.c -- Motif based terminal emulator

	fcterm.h &
	term.c -- Common subroutines for the two terminal emulators.

	[fcterm is 'Foxys Color Terminal emulator']
	[fmcterm is 'Foxys Motif Color Terminal emulator']

I  like  xterm  a  lot.  fcterm demonstrates how easy it is to
build   a   terminal   emulator   by   using  'clean'  calling
conventions.  In  addition fcterm contains the features I most
often  want  from xterm. (Theres no point in building a subset
of xterm if users frequently use xterm instead of fcterm).

fcterm and ctw were created in my spare time. This means that the
code may not be bug free, and certain features are incomplete (see
below). These will eventually be addressed and any person wishing
to contribute code, please feel free to do so.

It would be nice at some time in the future to submit ctw & fcterm
to the X consortium for inclusion in the core part of X11 to get over
the hurdle of xterm.

The ctw widget is configurable to emulate certain variants of ANSI
terminal emulators, specifically Interactive Unix console driver,
SCO Unix console driver, and xterm compatable functionality.
Many terminal emulators have small discrepancies in the way they 
handle certain features. For example, if the background color is set
to RED, and the ESC[K sequence is sent to erase to end of line, some
emulators erase to end of line using black, whilst others use the
current background color. Obviously the code to implement these two
is extremely tiny so having these types of things as attributes which
can be changed by the user or application is beneficial. The main
purpose in adding this sort of functionality is that I wanted ONE
program which could be used for testing software I write to make sure
it works in these different environments without the need to fire
up my 386 Unix system.

CTW contains some very useful facilities which I use a lot of the time.
Two of these are 'literal-mode' and 'slow-mode'. These modes are
using when I am testing my CRISP editor to see how well it performs
at slow baud rates. In literal mode, escape sequences, TABs, and \r
sequences are not interpreted but are printed in a readable fashion so
the user can see exactly what characters are being printed. In slow mode
ctw acts somewhat like a 1200-2400 baud terminal. This allows the user
to see where the cursor is going and detect any performance silliness.

CTW has been specifically optimised. It appears to run under most
circumstances as fast as xterm or even faster (see bugs section).
Certain things are pretty awful in terms of performance and these may
be looked into. For 99% of what you may do it is adequately fast enough.

CTW also supports mouse reporting escape sequences. At present
this allows the user to track ButtonPress, ButtonRelease, and Motion
events for the mouse -- selectable with an escape sequence. (ESC[?1962h
and ESC[?1962l). Please refer to the CTW_APPL_MOUSE attribute described
below for some useful comments on how this works.

The CTW widget
==============
The ctw widget creates a window which represents a character based
display. Each character is represented by a 4-byte structure to encode
color and other attribute information.

Data is written to the window using a public method function. This
function is called ctw_add_string():

	void ctw_add_string(Widget, char *str, int len);

'str' is the sequence of characters. 'len' says how long the
string is (this allows embedded NULLs -- which are ignored). We
do things this way because a real terminal emulator may have nulls
to print.

For example, to print 'hello world' to the screen:

	ctw_add_string(w, "Hello world", 10);

To print 'Hello world' with the 'hello' in one color and the 'world'
in another, do something like:

	ctw_add_string(w, "\033[32mHello \033[34mworld", 19);

In addition, there are other public functions which have been
written so that the fcterm terminal emulator could be written cleanly
without any real knowledge of how the widget is implemented:

ctw_get_top_line
================
The ctw widget is designed so that the application can have a scrollable
window (a la xterm). Lines which are scrolled off the screen are saved
so that they can be brought into view. The 'screen' may be thought of
as an array of lines, most of which are not visible. Normally there
are 'saveLines' off the top of the screen plus the 'rows' number of
lines visible. (See description of resources).


This function returns the line number at the top of the screen.
Given this number and the saveLines + rows resources, the application can
figure out how to draw the thumb part of a scrollbar.

ctw_set_top_line
================
This function would typically be called because the user is using
the scrollbar to look at the lines which have scrolled off the screen.

ctw_get_attributes & ctw_set_attributes
=======================================
Attributes are things like the current state of the autowrap flag, or
the ability to interpret certain escape sequences. Attributes are
represented as an array of integers. This function retrieves a
POINTER to the internal attributes associated with the widget. This
array should then be copied, modified locally and then set using the
ctw_set_attributes.

DO NOT MODIFY THE ARRAY OF ATTRIBUTES POINTED TO DIRECTLY.

ctw_get_selection
=================
This function is provided to allow the ctw widget to ask the X server
for the current copy of the Primary selection buffer. It is used
to implement the 'paste' semantic in the terminal emulator.

ctw_get_geometry
================
This function is used to retrieve the current dimensions of the
terminal window in rows & columns.

ctw_get_line
============
This function returns a pointer to an internal line structure
(an array of vbyte_t elements) representing the characters and attributes
on that line. This allows the application to read whats on the screen and
is used by fcterm/fmcterm to implement the Snap History option to dump the
possibly scrolled off screen contents to a file.

	----------------------------------
CTW Attributes
==============
The attributes provided by ctw_get_attributes() and set via
ctw_set_attributes() are integer valued. Attributes are defined
in the ctw.h include file as a set of enums of the form: CTW_xyz.

Some attributes are designed to be read-only. Most of the attributes are
effected by the standard escape sequences. (See top of ctw.c for a list
of sequences supported. This list should be up to date).

CTW_AUTOWRAP	Says whether cursor wraps around after right margin is hit.

CTW_AUTOLINEFEED
		Not used currently.

CTW_APPL_KEYPAD	
		Indicates whether keypad generates 'normal' keys or
		application escape sequences. Set via the ESC=/ESC>
		escapes sequences.

CTW_APPL_MOUSE	This attribute is turned on/off by the ESC[?1962h and
		ESC[?1962l escape sequence (or by the user from the
		terminal attributes popup dialog box). If mouse 
		reporting is enabled then mouse events cause escape sequences
		to be sent to the application (for input to the pty)
		reporting the current state and position of the mouse.

		At present there is a conflict between the user wanting
		to use the mouse for popups etc and the underlying
		application getting them. Its up to the application writer
		to decide how to arbitrate this.

		Please note that with the default translations, button
		1 will NOT be reported -- since it is mapped to the 
		hilighted selection action procedure. You will have
		to change the translation table in your code or in
		ctw.c to gain access to this.

		Refer to comment at the top of ctw.c for a description
		of the information reported to the application.

CTW_CURSOR_KEYPAD
		Specifies whether cursor keys generate ESC[x or ESCOx.
		Affected by the ESC[?1h and ESC[?1l escape sequences.

CTW_COLOR_RESET
		Specifies whether ESC[0m (reset attributes) resets
		the foreground/background colors or just leaves them alone.
		(SCO Unix and ISC Unix differ on this).

CTW_CURRATTR	Returns current attribute settings as follows:

		1111110000000000
		5432109876543210
		|--||--|    FBRU   flash,bold,reverse,underline
 		fg  bg            color numbers

		(It is not possible at present to set the attributes this
		way).

CTW_CUT_NEWLINES
		When hilighting text, specifies whether newlines should
		be inserted at newline boundaries. This attribute shouldn't
		really exist.

CTW_DESTRUCTIVE_TABS
		Says whether tabs generate spaces clearing any text they
		move over, or whether the cursor is simply moved to the next
		tab stop without destroying the intermediate characters.


CTW_ERASE_BLACK	Not currently used. Specifies whether things like
		ESC[K erase using white on a black background or using the
		current color settings.

CTW_ESC_LITERAL	Says whether ESC ESC should print an ESC character.
		(The ESC character is a >= in the fixed fonts). Needed
		for compatability with ISC/SCO console drivers.

CTW_FONT_SIZE	This is a READ-ONLY attribute. The high 16 bits are
		the font width. The low 16-bits are the font height.

CTW_LITERAL_MODE
		Turns on mode where escape sequences are not actioned but
		are printed literally. Useful for debugging applications.

CTW_LOGGING	If set to true then all data sent to the widget is logged
		in a log file (set by the 'logFile' resource). The log file
		name can be changed with the ESC]46;log-file<BEL>
		escape sequence.

CTW_NEWLINE_GLITCH
		Specifies whether the DEC/xterm 'feature' of causing the
		cursor to hang around in the right hand column to be 
		emulated. In this mode, the cursor sticks to the right hand
		column until the next printable character is received and
		avoids scrolling until absolutely necessary. 

CTW_PC_CHARSET	Used to emulate the PC character set. At present only the
		line drawing characters are remapped.

CTW_RESET	A single shot attribute. Setting this attribute is similar
		to sending the ESCc escape sequence and is used to reset
		all the screen attributes. It is set back to zero immediately
		it is set.

CTW_SCROLLING_REGION
		This is a read-only flag. It is set to TRUE whilst a
		scrolling region is in effect (ESC[x;yr). Used internally
		but also useful in fcterm to tell whether an application has
		accidentally left the screen in scrolling region mode.
		(Can be very difficult to tell otherwise).

CTW_SIZE	This is a read-only attribute. The top 16 bits contains
		the number of columns, and the bottom 16 contains the number
		of rows. Useful for informtional purposes only. (These
		values can be gotton from the resources as well).

CTW_SPEED	This attribute controls how fast the display is updated.
		It is a value which is a percentage and is normally set to
		100% (full-speed). Lower values will slow down the display.
		This is designed for watching fast screen updates on a fast
		machine.

CTW_SMOOTH_SCROLL
		Same as xterm feature. If set then scrolling occurs one
		line at a time. If false, then ctw looks ahead and tries to
		scroll up to a whole screens worth for maximum performance.

CTW_SUN_FUNCTION_KEYS
		Says whether function keys generate Sun compatable
		escape sequences or VT100 sequences. This attribute
		is synonymous with the 'sunFunctionKeys' resource.

CTW_ISC_DRIVER	This flag acts as a macro. When set, then other attributes
		are set so that ctw emulates an Interactive Unix console
		driver.

CTW_SCO_DRIVER	This flag acts as a macro. When set, then other attributes
		are set so that ctw emulates an SCO Unix console driver.

Action Event Functions
======================
focus_change(n)
		'n' should be 0 or 1. This function can be placed in the
		translation table to allow the cursor flashing to stop when
		the appropriate event happens. '0' is used to stop flashing.
		'1' is used to cause flashing. This procedure would normally
		be called on an <EnterWindow>/<LeaveWindow> or 
		<FocusIn>/<FocusOut> event transition.

Callbacks
=========
The ctw widget supports various callbacks, all accessible as resources.
All callbacks pass the following structure to the callback routine:

typedef struct ctw_callback_t {
	enum ctw_reasons	reason;
	char			*ptr;
	int			len;
	int			top_line;
	XEvent			*event;
	} ctw_callback_t;

Not all fields are valid for all callbacks.

applCallback	Callback called when certain escape sequences are parsed
		to do with the application. Typically this is to iconise
		the application or change its size. These sequences cannot
		be handled solely by the widget because some of them affect
		the application shell widget or any underlying pty.

		reason = CTW_OPEN_WINDOW
			Uniconise the application shell widget.

		reason = CTW_CLOSE_WINDOW
			Iconise the application shell widget.

		reason = CTW_MOVE_WINDOW
		x = pixel co-ordinate of left margin
		y = pixel co-ordinate of top margin
			Move window to specified co-ordinates

		reason = CTW_SIZE_WINDOW_PIXELS
		width = width in pixels of new window size.
		height = height in pixels of new window size.
			Resize application shell.

		reason = CTW_FRONT_WINDOW
			Raise window to top of stack.

		reason = CTW_BACK_WINDOW
			Lower window to bottom of stack.

		reason = CTW_SIZE_WINDOW_CHARS
		width = new width in characters
		height = new height in characters
			Resize window specifying size in characters.

color1
color2
 ...
color16		By default the CTW widget uses 16 colors corresponding
		to the definitions specified by the ANSI/ISO standard.
		On a monochrome screen gray scales are used which may not
		be appropriate. These attributes allow the user or
		appliaction to change the definition of colors, where
		color1..color8 are the foreground colors, and color9..
		color16 are the background colors. Standard X color
		names should be used.

kbdCallback	Called for keyboard input.

		reason = CTW_INPUT
		ptr = pointer to buffer containing characters typed
		len = number of bytes in buffer.

resizeCallback	Called when widget is resized

		reason = CTW_RESIZE

topCallback	Called when the line at the top of the screen changes.
		This callback is used so that a scrollbar can place the thumb
		at a proportional point to where the 'top-line' is in terms
		of the lines scrolled off the screen.

		reason = CTW_TOP_LINE
		top_line = new line number

mouseCallabck	Called when button activity is detected.

		reason = CTW_SELECTION
			called when the select-start action routine is
			invoked. This happens when user normally hits
			button1 to start hilighting a region.

		reason = CTW_BUTTON_DOWN
		event = X event
			called when a button down event is detected.

		reason = CTW_BUTTON_MOTION
		event = X event
			called when a motion event is detected whilst a
			mouse button is held down.

		reason = CTW_BUTTON_UP
		event = X event
			called when a button up event is detected.

Resources
=========
Many of the widget resources are compatable with those in xterm. ctw
doesnt implement all of the xterm ones. Maybe it should.

In the table below, the resource name only is listed. Names follow standard
conventions. To use these resources in your code, just prepend
resource names with XtN or XtC for class names (capitalize first letter).
For example, XtNfont or XtCFont, etc.

font		Font used within window.

cursorColor	Color of cursor. Any X color can be used. (Colors used in
		the widget window are limited to the 16 standard colors
		of a color terminal).

rows		Define number of visible rows in the window.
columns		Define width of window in characters.

kbdCallback	Callback used to handle keyboard input.

sunFunctionKeys	Xterm compatible resource specifying what escape sequences
		should be generated by default for function keys.
		(Translation tables may be used to override this).

saveLines	Number of lines saved when they scroll off the top of the
		window.

topCallback	Callback used for handling scrollbars.

mouseCallback	Callback for button events and selection handling.

geometry	Not really sure! Should be used so that we can interpret
		command line geometry requests using characters sizes
		(not pixel sizes), but doesnt currently work.

flashrate	Time (in mSec) between cursor flashes. Cursor only
		flashes while we have the input focus.

font1, font2, font3, font4, font5, font6
		Not used by the ctw widget, but used by fcterm. Names
		are chosen to be xterm compliant. These fonts define
		the entries in the Change Fonts menu.

hiliteBackground
hiliteForeground
		Colors to use when hilighting text for the Primary selection.

multiClickTime	Time (in mSec) used for detecting double/triple mouse clicks.

internalHeight	Specifies the amount of room to leave at the top and bottom
		of the window.
internalWidth	Specifies the amount of room to leave at the sides of the
		window.

logFile		Name of log file to send output. (Default CtwLog.XXXXXX).
logging		When set to TRUE all output to the window is sent to the
		logfile.

The fcterm Terminal emulator
============================
fcterm is designed to be a plug-in replacement for xterm. It
accepts many of the same resource and class definitions as xterm.
The user interface is somewhat different, but it implements most of the
features of xterm that I use. (There is NO Tektronix support. Use
xterm if you need access to this facility).

In normal use, fcterm and xterm operate almost identically. The main
difference is in the button bindings.

Button1 is used for hilighting text (same as xterm although only
single and triple clicking are currently supported).

Button2 is used for the popup menu at present. Its functionality
will change at a later date. It may be used in future to be compatable
with the xterm 'extend-selection' functionality or it may be used for
some other use.

Button3 pops up a menu. The menu structure is not the same as xterm.
Currently, button3 has the following entries in the menu:

Toggle scrollbar 	turn on/off scrollbar. Something I do
			very frequently.
Reset character set	Used when terminal is in line-drawing character
			set but it shouldnt be!
Paste			Paste Primary selection. Hopefully avoids the
			accidental and infuriating feature of xterm where
			you accidentally hit the 'stuff' mouse button.
Terminal attributes...	Pops up a dialog box to allow the CTW widget
			attributes to be changed/examined.
Status monitor...	Pops up a little window showing the number of
			characters received, number chars/sec processed.
			Useful for performance monitoring.
Change font...		Used to make a font selection

Bugs & Missing Features
=======================
ctw:
----
Internal width & height are not processed properly.


Changes
Sat Dec 30 22:05:39 2017  fox

     180* include/ctw.h
          src/ctw.c
          src/fcterm.c
          fix time_str().

Thu Feb 19 23:12:31 2015  fox

     179* .gitignore
          src/ctw.c
          Make 0xa0 visible - really annoying.

Wed Feb 18 22:42:38 2015  fox

     178* include/ctw.h
          include/ctwP.h
          src/ctw.c
          Add <space>/d/f/p/u support in ctrl-o mode.

Mon Feb 16 22:40:27 2015  fox

     177* include/ctwP.h
          src/ctw.c
          src/usage.c
        Add support for Ctrl-O flushing.

     176* include/ctwP.h
          makefile
          src/ctw.c
          src/term.c
          src/usage.c
          src/usage.txt
          Add dynamic words as per #175.

     175* src/ctw.c
          Add support for undefined/uninitialized and load words
	  from CTW_HIGHLIGHT_WORDFILE.

Thu Sep 11 21:24:57 2014  fox

     174* src/fcterm.c
          src/group.c
          Do synchronous call to XConfigureWindow.

Wed Sep 10 22:54:06 2014  fox

     173* include/fcterm.h
          makefile
          src/fcterm.c
          src/group.c
          Add 'fcterm restore' support.

Mon Sep  1 20:34:31 2014  fox

     172* include/fcterm.h
          src/fcterm.c
          src/term.c
          Fix 'fcterm -ls' not working.

Thu Aug 28 22:28:12 2014  fox

     171* include/fcterm.h
          include/sequence.h
          makefile
          mkseq.sh
          src/ctw.c
          src/fcterm.c
          src/term.c
          Add 'fcterm create-screen'.
	  Allow type in whilst in minimap mode.

Wed Aug 27 08:48:04 2014  fox

     170* include/fcterm.h
          makefile
          src/ctw.c
          src/fcterm.c
          src/group.c
          Launch group_init after fork, not before.

     169* archive.sh
          src/group.c
          Archive: 1.0021

     168* include/ctw.h
          include/ctwP.h
          include/fcterm.h
          include/sequence.h
          machines/linux-x86_64.sh
          makefile
          mkseq.sh
          src/ctw.c
          src/fcterm.c
          src/group.c
          src/term.c
          Added group support and the "fcterm <cmd>" subcommands.

Tue Aug 12 23:33:45 2014  fox

     167* include/ctw.h
          include/fcterm.h
          src/ctw.c
          src/fcterm.c
          Add "Search" to the popup menu.

Sun Jul  6 22:10:26 2014  fox

     166* include/ctw.h
          include/sequence.h
          mkseq.sh
          samples/24-bit-color.pl
          src/ctw.c
          src/ptysrv.c
          src/term.c
          Add ESC[38;2;rr;gg;bb;m and ESC[48;2;rr;gg;bb;m support

Thu Jul  3 21:25:41 2014  fox

     165* src/term.c
          More aggressive backing off in cb_read to avoid uncontrol ^C.

Wed Jul  2 22:33:44 2014  fox

     164* makefile
          src/term.c
          Fix the infinite output/Ctrl-C problem by terminating cb_read
	  if any X11 input.

Sat May 31 13:37:52 2014  fox

     163* 
          include/ctw.h
          include/ctwP.h
          src/ctw.c
          Build with stpcpy to workaround gcc/glibc bug.

Fri May 23 23:05:11 2014  fox

     162* src/ctw.c
          Add hyperlink detection/display.

Sat Jan 25 21:20:03 2014  fox

     161* README, makefile, src/fcterm.c, src/scrbar.c:
        Make the background panel less bright to make it clearer.

Tue Aug 27 22:39:16 2013  fox

     160* armv6l: binary supported

Tue Dec 18 23:01:08 2012  fox

     159* ctw.c: Fix cursor positioning for special line drawing (gridlines).

Mon Oct 29 22:58:58 2012  fox

     158* build.h: 4.34

     157* ctw.c: Add support for ESC[? p1;p2;p3 c Linux cursors.

Sun Sep  9 09:22:02 2012  fox

     156* build.h: 4.33

     155* ctw.c: read_modifier_keys: Avoid redoing multiple times.

Sat Aug 25 17:37:00 2012  fox

     154* ctw.c: Add support for ESC[9Xm and ESC[10Xm for eastman ball.

Mon Aug  6 12:17:54 2012  fox

     153* ctw.c: Allow KP_0 to be an INS key innon-keypad mode.

Sun Jul  8 18:23:15 2012  fox

     152* term.c, makefile, fcterm.c, ctw.c: Build on MacOS 10.7

Thu Jun 21 22:07:15 2012  fox

     151* ctw.c: Add UTF-8 0x2500 line drawing chars for elinks.

Thu May  3 21:18:02 2012  fox

     150* term.c: Add _NET_WM_ICON support.

Sat Oct 22 20:40:18 2011  fox

     149* ctw.c, fcterm.c: Add shell status "LED" to the status panel.

Thu Oct 20 22:48:18 2011  fox

     148* ctw.c: Add support for collapse/open graphic.

Wed Oct 19 21:11:00 2011  fox

     147* ctw.c: Fix shifted/line drawing problem.

Tue Oct 18 23:09:44 2011  fox

     146* ctw.c: Better outline grid drawing.

Mon Oct 17 22:37:06 2011  fox

     145* ctw.c: Support 0x01 gridlines drawing for CRiSP.

Wed Oct 12 22:10:57 2011  fox

     144* ctw.c: Fix issue in 142 - we didnt create the log dir for the user.

Sun Oct  2 21:37:24 2011  fox

     143* ctw.c: Add support for 1936/1937 box/unbox drawing.

Mon Sep 26 23:18:40 2011  fox

     142* term.c: Put logs into /tmp/$USER dir.

Tue Sep 13 22:13:12 2011  fox

     141* ctw.c: Allow hilite/select of a single space.

Sat Aug 27 15:24:26 2011  fox

     140* ctw.c: Optimise DRAW_SET_FOREGROUND and DRAW_SET_BACKGROUND.

Fri Aug 26 22:18:50 2011  fox

     139* ctw.c, mkseq.sh, term.c: Add CTW_PID env var. Add ESC[1935m to
          dump a status file so we can read the widnow size and font info
	  without playing with stdin of an app.

Tue Aug 23 21:15:02 2011  fox

     138* ctw.c: Continuing fix for dangling esc sequences.

Mon Aug 22 20:26:05 2011  fox

     137* ctw.c: Fix problem with continued escape sequences with
          text fields.

     136* ctw.c: Avoid core dump when we cannot find a font.

Wed Aug 10 18:15:55 2011  fox

     135* ctw.c: Dont core dump when drawing graphical label and no font
          specified.

     134* term.c: Add CTW_TERM=1 to the environment so children know
          we are running inside fcterm.

Mon Aug  1 21:15:53 2011  fox

     133* ctw.c: Display NUL bytes however.

     132* term.c: Dont interpret NUL special when not using ptysrv.

Mon Jul 25 20:02:11 2011  fox


mkseq.sh
sed -e 's/^/"/g' -e 's/$/\\r\\n"/g' <<EOF

^N x ^O -- switch char set / line drawing
   x == 0x01    Mid-char vertical line

Escape sequences:
   ESC ESC x  Print character x literally.
   ESC 7	Save cursor position
   ESC 8	Restore cursor position
   ESC @0	Not in curses.
   ESC @1     In curses emulation mode.
   ESC @2     Receive ioctl.
   ESC @3     Reply to ioctl.
   ESC c	Reset terminal
   ESC D	Index (down cursor or scroll).
   ESC H	Set tab stop at current position.
   ESC M	Reverse index (up cursor or insline).
   ESC Z	Identify terminal.
   ESC >	Numeric keypad
   ESC =	Application keypad
   ^H		Backspace
   ^I		Tabs (fixed-width -- 8 chars wide).
   ^J		Line feed
   ^L		Clear screen.
   ^M 	Carriage return
   ESC(0   ESC)0   ESC*0   ESC+0  Set character sets.
              0  Normal line drawing font set
              1  Alternate fonts
              2  Alternate fonts
              A  American ASCII
              B  British ASCII
   ESC [n @	Insert characters.
   ESC [n A	Move cursor up
   ESC [n B	Move cursor down
   ESC [n C	Move cursor right
   ESC [n D	Move cursor left
   ESC [n d	Goto (x, n)
   ESC [n E	Goto (0,y+n)
   ESC [n F	Goto (0,y-n)
   ESC [n G	Goto specified column.
   ESC [n;m H Move cursor to absolute position.
   ESC [n J   Erase screen 0==clear below; 1==clear above; 2==clear all
   ESC [n L	Insert lines at current row
   ESC [n M	Delete lines at current row
   ESC [n P	Delete characters.
   ESC [S	Index (down cursor or scroll).
   ESC [T	Reverse index (up cursor or insline).
   ESC [n X	Erase to blank (no cursor move)
   ESC [n Y	Set mouse button mask.
***ESC [n Z	Request mouse position.
   ESC [Z  	Backtab

   ESC [ c	Ask for answer back message.
   ESC [n;m f Move cursor to absolute position.
   ESC [n g	Print ASCII char 'n' (SCO mode only)
   ESC [3 g	Clear TAB stops.
   ESC [4 h	Enter insert mode
   ESC [4 l	Exit insert mode
   ESC [0 m	Reset attributes.
   ESC [1 m	Bold
   ESC [4 m	Underline
   ESC [5 m	Blink
   ESC [7 m	Reverse video.
   ESC [8 m	Store current colors as the default.
   ESC [10 m  Select primary font, don't display ctrl chars, do't set bit 8
   ESC [11 m  Select first alternate font; lets chars < 32 be displayed
   ESC [21m	Reset bold.
   ESC [24m	Reset underline.
   ESC [25m	Reset blink.
   ESC [27m	Reset reverse video.
   ESC [3Xm	Set foreground color.
   ESC [38 ; 2 ; r ; g ; b m	Set foreground RGB.
   ESC [38 ; 5 ; XX m	Set foreground from 256 palette.
   ESC [39m	Restore foreground default color.
   ESC [4Xm	Set background color.
   ESC [48 ; 2 ; r ; g ; b m	Set foreground RGB.
   ESC [48 ; 5 ; XX m	Set background from 256 palette.
   ESC [49m	Restore background default color.
   ESC [7Xm	Set cursor color.
   ESC [78m	Restore cursor color.
   ESC [1900 m  Stop literal mode.
   ESC [1901 m  Go into literal mode.
   ESC [1902 m  Go into strikethru mode.
   ESC [1903 m  Exit strikethru mode.
   ESC [1904 m  Enter autoswitch mode.
   ESC [1905 m  Exit autoswitch mode.
   ESC [1920 ; x ; y ; w ; h ; m         Draw a rectangle.
   ESC [1921 ; x ; y ; w ; h ; m         Fill a rectangle.
   ESC [1922 ; x1 ; y1; x2 ; y2; m       Draw a line
   ESC [1923 ; x ; y ; m                 Draw a pixel.
   ESC [1924 m Clear graphics chain.
   ESC [1925 m Clear graphics chain and redraw screen.
   ESC [1926 ; x; y; w; h; name m        Set name
   ESC [1927 ; x; y; str; m              Draw string
   ESC [1928 ; x; y; w; h; arc1; arc2; m Arc
   ESC [1929 ; x; y; w; h; arc1; arc2; m FillArc
   ESC [1930 ; rgb m                     Set background
   ESC [1931 ; rgb m                     Set foreground
   ESC [1932 ; name ; m                  Set font
   ESC [1933 ; x; y; str; m              Draw image string
   ESC [1934 m                           Query winsize in pixels (wxh<Enter>)
   ESC [1935 m                           Dump internal status to a /tmp/fcterm/fcterm.$CTW_PID
   ESC [1936 m    Box each character drawn.
   ESC [1937 m    Unox each character drawn.
   ESC [1938 ; n m  Group fcterms (n == 1 - group; == 0 - ungroup)
   ESC [1939 m   Show status.
   ESC [1940 m   Search/command mode.
   ESC [1941 m   Zoom
   ESC [1942 m   Minimap
   ESC [1943; <id> m   Create new screen.  
   ESC [n;m r	Set scrolling region to lines n..m.
   ESC [n;m r	Set scrolling region.
   ESC [s	Saved cursor position.
   ESC [n t	Shelltool/cmdtool compatable escape sequence
   		ESC [1t	Open window.
   		ESC [2t	Iconise window.
   		ESC [3t	Move window.
   		ESC [4t	Change size (in pixels) window.
   		ESC [5t	Raise window.
   		ESC [6t	Lower window.
   		ESC [8t	Change size (in chars) window.
   ESC [u	Restore cursor position.

   ESC [ ? 1 h	Application cursor keys
   ESC [ ? 3 h	80/132 column mode switch.
   ESC [ ? 4 h	Set smooth scroll mode.
   ESC [ ? 5 h	Reverse video mode.
   ESC [ ? 7 h	Set autowrap
   ESC [ ? 8 h	Enable autorepeat (no-op) Use XAutoRepeatOn)
   ESC [ ? 9 h	Turn on application mouse.
   ESC [ ? 12 h	Move this console to the front.
   ESC [ ? 25 h	Turn on cursor.
   ESC [ ? 47 h	Use alternate screen buffer.
   ESC [ ? 1000 h	Enter keypad mode.
   ESC [ ? 1034 h	SMM mode
   ESC [ ? 1049 h	Use alternate screen buffer and screen clear.

   ESC [ ? 1 l	Normal cursor keys
   ESC [ ? 3 l	80/132 column mode switch.
   ESC [ ? 4 l	Set jump scroll mode.
   ESC [ ? 5 l	Normal video mode.
   ESC [ ? 7 l	Disable autowrap
   ESC [ ? 8 l	Disable autorepeat (no-op) Use XAutoRepeatOn)
   ESC [ ? 9 l        Turn off application mouse.
   ESC [ ? 25 l	Turn off cursor.
   ESC [ ? 47 l	Use normal screen buffer.
   ESC [ ? 1000 l	Exit  keypad mode.
   ESC [ ? 1049 l	Use alternate screen buffer/save cursor/scr clear

   ESC [ 6 n          Ask terminal for rows,col - ESC[r;cR
   ESC [ ? 3 r	Disable window resize.
   ESC [ ? 4 r	Enable window resize.

   ESC [ ? r;c;h;w;n S Scroll rectangular region up or down n lines

   ESC ] 0;<window/icon-title><BEL>
   ESC ] 1;<icon name><BEL>
   ESC ] 2;<window name><BEL>
   ESC ] 46;<log-file><BEL>    Change log file name.
   ESC ] 50;<font-name><BEL>   Change font name.
   ESC ] 100;<mini-label><BEL> Change label on status line.

   ESC [ ? 0 f	Enable Graphics Expose processing.
   ESC [ ? 0 F	Disable Graphics Expose processing.
   ESC [ ? 1 f	Toggle fg/bg color
   ESC [ ? 2 f	Push current font onto stack.
   ESC [ ? 3 f	Pop font from stack.

   ESC [ = C    	Hide cursor
   ESC [ = 1C    	Invisible cursor.

   ESC # 3    	Top-half of double height
   ESC # 4    	Bottom-half of double height
   ESC # 5    	Normal height
   ESC # 6    	Double width

   ESC [ > 0 c	Device attributes request. (secondary)

   Sequences reported:
     ESC [ b; m; s; r; c M  Mouse position report
   			b == Button being reported (1..3)
   			s == 0 button pressed
   			  == 1 button released
   			  == 2 button motion

   			m == Modifier status as a set of bits
   			     0x01 Shift down
   			     0x02 Ctrl down
   			     0x04 Meta down

   			r == current row (decimal 0..n)
   			c == current col (decimal 0..n)

Linux Console private sequences:

       ESC [ 1 ; n ]       Set color n as the underline color
       ESC [ 2 ; n ]       Set color n as the dim color
       ESC [ 8 ]           Make the current color pair the default attributes.
       ESC [ 9 ; n ]       Set screen blank timeout to n minutes.
       ESC [ 10 ; n ]      Set bell frequency in Hz.
       ESC [ 11 ; n ]      Set bell duration in msec.
       ESC [ 12 ; n ]      Bring specified console to the front.
       ESC [ 13 ]          Unblank the screen.
       ESC [ 14 ; n ]      Set the VESA powerdown interval in minutes.

       ESC [ ? p1 ; p2 ; p3 c Set cursor style
	    p1=	0  Don't change
		1  No cursor
		2  Underline
		3  Lower third of the block
		4  Lower half of the block
		5  Two thirds of the block
		6  Block cursor
	    p2/p3 = 0xbf = background/foreground color
EOF
